waku/common/sqlite/database.nim

```diff
@@ -0,0 +1,342 @@
+  discard sqlite3_reset(s) # same return information as step
+  discard sqlite3_clear_bindings(s) # no errors possible
+
+  res
+
+template readResult(s: RawStmtPtr, column: cint, T: type): auto =
+  when T is Option:
+    if sqlite3_column_type(s, column) == SQLITE_NULL:
+      none(typeof(default(T).get()))
+    else:
+      some(readSimpleResult(s, column, typeof(default(T).get())))
+  else:
+    readSimpleResult(s, column, T)
+
+template readResult(s: RawStmtPtr, T: type): auto =
+  when T is tuple:
+    var res: T
+    var i = cint 0
+    for field in fields(res):
+      field = readResult(s, i, typeof(field))
+      inc i
+    res
+  else:
+    readResult(s, 0.cint, T)
+
+proc exec*[Params, Res](s: SqliteStmt[Params, Res],
+                        params: Params,
+                        onData: DataProc): DatabaseResult[bool] =
+  let s = RawStmtPtr s
+  bindParams(s, params)
+
+  try:
+    var gotResults = false
+    while true:
+      let v = sqlite3_step(s)
+      case v
+      of SQLITE_ROW:
+        onData(s)
+        gotResults = true
+      of SQLITE_DONE:
+        break
+      else:
+        return err($sqlite3_errstr(v))
+    return ok gotResults
+  finally:
+    # release implicit transaction
+    discard sqlite3_reset(s) # same return information as step
+    discard sqlite3_clear_bindings(s) # no errors possible
+
+
+proc query*(db: SqliteDatabase, query: string, onData: DataProc): DatabaseResult[bool] =
+  var s = prepare(db.env, query): discard
+
+  try:
+    var gotResults = false
+    while true:
+      let v = sqlite3_step(s)
+      case v
+      of SQLITE_ROW:
+        onData(s)
+        gotResults = true
+      of SQLITE_DONE:
+        break
+      else:
+        return err($sqlite3_errstr(v))
+    return ok gotResults
+  finally:
+    # release implicit transaction
+    discard sqlite3_reset(s) # same return information as step
+    discard sqlite3_clear_bindings(s) # no errors possible
+    discard sqlite3_finalize(s) # NB: dispose of the prepared query statement and free associated memory
+
+proc prepareStmt*(
+  db: SqliteDatabase,
+  stmt: string,
+  Params: type,
+  Res: type
+): DatabaseResult[SqliteStmt[Params, Res]] =
+  var s: RawStmtPtr
+  checkErr sqlite3_prepare_v2(db.env, stmt, stmt.len.cint, addr s, nil)
+  ok SqliteStmt[Params, Res](s)
+
+proc close*(db: SqliteDatabase) =
+  discard sqlite3_close(db.env)
+
+  db[] = SqliteDatabase()[]
+
+
+##  Maintenance procedures
+
+# TODO: Cache this value in the SqliteDatabase object.
+#       Page size should not change during the node execution time
+proc getPageSize*(db: SqliteDatabase): DatabaseResult[int64] =
+  ## Query or set the page size of the database. The page size must be a power of 
+  ## two between 512 and 65536 inclusive.
+  var size: int64
+  proc handler(s: RawStmtPtr) = 
+    size = sqlite3_column_int64(s, 0)
+
+  let res = db.query("PRAGMA page_size;", handler)
+  if res.isErr():
+      return err("failed to get page_size")
+
+  ok(size)
```

Comment:
Is there a format for returning results? Maybe an addition to the style guide - `ok` vs `return ok`